### PR TITLE
Moved some plain Vulkan code

### PR DESCRIPTION
**Pull request by shurunxuan**
This PR moves a lot of plain Vulkan code into some encapsulated classes to provide some easy-to-use functionalities including:

* Instance & Device creation
  * Layer and Extension enumeration
* Swapchain creation & recreation
  * Surface included
* Command Pool and Command Buffer management
  * Automatically creates command pools per frame per thread, and manages the command buffers created
  * Allows reuse of command buffers by just resetting the command pool at the start of the frame 
* Debug messengers and Object Debug Names
  * ~~For now this will crash the App when launched by RenderDoc~~ Fixed by [4eb247c](https://github.com/gore-engine/gore/pull/29/commits/4eb247c7588c9e15e68380671bdd9839bd77224c), see [VUID-VkDebugUtilsMessengerCallbackDataEXT-pMessageIdName-parameter](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDebugUtilsMessengerCallbackDataEXT.html#VUID-VkDebugUtilsMessengerCallbackDataEXT-pMessageIdName-parameter), `pMessageIdName` is possible to be `NULL`

### CODE REVIEW

File: src/Graphics/CommandPool.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    CommandPool& operator=(CommandPool&& other) noexcept;
+
+    [[nodiscard]] const vk::raii::CommandPool& Get(uint32_t swapchainImageIndex);
+
+    [[nodiscard]] const vk::raii::CommandBuffer& GetCommandBuffer(uint32_t swapchainImageIndex);
+    void Reset(uint32_t swapchainImageIndex);
+
+private:
+    struct Pool
```
Comment (Top61ly) [inline src/Graphics/CommandPool.h]: I'm using class Pool and class Handle to write graphics resource system. Could you please change the name of Pool even it is a private struct?

**Review by Top61ly — CHANGES_REQUESTED** [review]
I'm using class Pool and class Handle to write graphics resource system. Could you please change the name of

File: src/Graphics/CommandPool.h (hunk 1 above)
Comment (shurunxuan) [reply src/Graphics/CommandPool.h]: Will do.

**Review by Top61ly — APPROVED** [review]
OK!

**Pull request merged**